diff --git a/internal/process/process_command.go b/internal/process/process_command.go
--- a/internal/process/process_command.go
+++ b/internal/process/process_command.go
@@ -300,6 +300,7 @@ func (p *ProcessCommand) run() {
 					cmdCancel = res.cancel
 					fn := res.handlerFn
 					p.handler.Store(&fn)
+					p.lastUse.Store(time.Now().UnixNano())
 					setState(StateReady)
 					notifyWaiters(nil)
 					if req.block {
